File: Tau/andstudio/Tau/app/src/main/java/cymheart/tau/effect/Compressor.java
package cymheart.tau.effect;

public class Compressor extends Effect {

    public Compressor()
    {
        ndkEffect = ndkCreateCompressor();
    }

    //设置采样频率
    public void SetSampleFreq(float freq)
    {
        ndkSetSampleFreq(ndkEffect, freq);
    }

    //设置Attack时长
    public void SetAttackSec(float sec) {
        ndkSetAttackSec(ndkEffect, sec);
    }

    //设置Release时长
    public void SetReleaseSec(float sec) {
        ndkSetReleaseSec(ndkEffect, sec);
    }

    //设置比值
    public void SetRadio(float radio) {
        ndkSetRadio(ndkEffect, radio);
    }

    //设置门限
    public void SetThreshold(float threshold) {
        ndkSetThreshold(ndkEffect, threshold);
    }

    //设置拐点的软硬
    public void SetKneeWidth(float width) {
        ndkSetKneeWidth(ndkEffect, width);
    }

    //设置增益补偿
    public void SetMakeupGain(float gain) {
        ndkSetMakeupGain(ndkEffect, gain);
    }

    //设置是否自动增益补偿
    public void SetAutoMakeupGain(boolean isAuto) {
        ndkSetAutoMakeupGain(ndkEffect, isAuto);
    }

    //计算系数
    public void CalculateCoefficients() {
        ndkCalculateCoefficients(ndkEffect);
    }

    private static native long ndkCreateCompressor();
    private static native void ndkSetSampleFreq(long ndkEffect, float freq);
    private static native void ndkSetAttackSec(long ndkEffect, float sec);
    private static native void ndkSetReleaseSec(long ndkEffect, float sec);
    private static native void ndkSetRadio(long ndkEffect, float radio);
    private static native void ndkSetThreshold(long ndkEffect, float threshold);
    private static native void ndkSetKneeWidth(long ndkEffect, float width);
    private static native void ndkSetMakeupGain(long ndkEffect,float gain);
    private static native void ndkSetAutoMakeupGain(long ndkEffect, boolean isAuto);
    private static native void ndkCalculateCoefficients(long ndkEffect);

}

File: Tau/src/core/FX/Compressor.cpp
#include"Compressor.h"

namespace tauFX
{
	Compressor::Compressor()
	{
		leftCompressor = new dsignal::Compressor();
		rightCompressor = new  dsignal::Compressor();
	}

	Compressor::~Compressor()
	{
		DEL(leftCompressor);
		DEL(rightCompressor);
	}

	void Compressor::Clear()
	{
		leftCompressor->Clear();
		rightCompressor->Clear();
	}

	//设置采样频率
	void Compressor::SetSampleFreq(float freq)
	{
		leftCompressor->SetSampleFreq(freq);
		rightCompressor->SetSampleFreq(freq);
	}

	//设置Attack时长
	void Compressor::SetAttackSec(float sec)
	{
		leftCompressor->SetAttackSec(sec);
		rightCompressor->SetAttackSec(sec);
	}

	//设置Release时长
	void Compressor::SetReleaseSec(float sec)
	{
		leftCompressor->SetReleaseSec(sec);
		rightCompressor->SetReleaseSec(sec);
	}

	//设置比值
	void Compressor::SetRadio(float radio)
	{
		leftCompressor->SetRadio(radio);
		rightCompressor->SetRadio(radio);
	}

	//设置门限
	void Compressor::SetThreshold(float threshold)
	{
		leftCompressor->SetThreshold(threshold);
		rightCompressor->SetThreshold(threshold);

	}

	//设置拐点的软硬
	void Compressor::SetKneeWidth(float width)
	{
		leftCompressor->SetKneeWidth(width);
		rightCompressor->SetKneeWidth(width);
	}

	//设置增益补偿
	void Compressor::SetMakeupGain(float gain)
	{
		leftCompressor->SetMakeupGain(gain);
		rightCompressor->SetMakeupGain(gain);
	}

	//设置是否自动增益补偿
	void Compressor::SetAutoMakeupGain(bool isAuto)
	{
		leftCompressor->SetAutoMakeupGain(isAuto);
		rightCompressor->SetAutoMakeupGain(isAuto);
	}

	//计算系数
	void Compressor::CalculateCoefficients()
	{
		leftCompressor->CalculateCoefficients();
		rightCompressor->CalculateCoefficients();
	}


	void Compressor::EffectProcess(float* leftChannelSamples, float* rightChannelSamples, int channelSampleCount)
	{
		for (int i = 0; i < channelSampleCount; i++)
		{
			leftChannelSamples[i] = leftCompressor->Process(leftChannelSamples[i]);
			rightChannelSamples[i] = rightCompressor->Process(rightChannelSamples[i]);
		}
	}

	void Compressor::EffectProcess(float* synthStream, int numChannels, int channelSampleCount)
	{
		if (numChannels == 2) {
			for (int i = 0; i < channelSampleCount * numChannels; i += 2)
			{
				synthStream[i] = leftCompressor->Process(synthStream[i]);
				synthStream[i + 1] = rightCompressor->Process(synthStream[i + 1]);
			}
		}
		else {
			for (int i = 0; i < channelSampleCount; i++)
			{
				synthStream[i] = leftCompressor->Process(synthStream[i]);
			}
		}
	}
}
